src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,678 @@
+(sdiff_le_iff (a b c : α) : a \ b ≤ c ↔ a ≤ b ⊔ c)
+(top_sdiff (a : α) : ⊤ \ a = ￢a)
+
+@[priority 100] -- See note [lower instance priority]
+instance generalized_heyting_algebra.to_order_top [generalized_heyting_algebra α] : order_top α :=
+{ ..‹generalized_heyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance generalized_coheyting_algebra.to_order_bot [generalized_coheyting_algebra α] :
+  order_bot α :=
+{ ..‹generalized_coheyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance heyting_algebra.to_bounded_order [heyting_algebra α] : bounded_order α :=
+{ ..‹heyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance coheyting_algebra.to_bounded_order [coheyting_algebra α] : bounded_order α :=
+{ ..‹coheyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance biheyting_algebra.to_coheyting_algebra [biheyting_algebra α] : coheyting_algebra α :=
+{ ..‹biheyting_algebra α› }
+
+/-- Construct a Heyting algebra from the Heyting implication alone. -/
+@[reducible] -- See note [reducible non instances]
+def heyting_algebra.of_himp [distrib_lattice α] [bounded_order α] (himp : α → α → α)
+  (le_himp_iff : ∀ a b c, a ≤ himp b c ↔ a ⊓ b ≤ c) : heyting_algebra α :=
+{ himp := himp,
+  compl := λ a, himp a ⊥,
+  le_himp_iff := le_himp_iff,
+  himp_bot := λ a, rfl,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a Heyting algebra from the Heyting implication alone. -/
+@[reducible] -- See note [reducible non instances]
+def heyting_algebra.of_compl [distrib_lattice α] [bounded_order α] (compl : α → α)
+  (le_himp_iff : ∀ a b c, a ≤ compl b ⊔ c ↔ a ⊓ b ≤ c) : heyting_algebra α :=
+{ himp := λ a, (⊔) (compl a),
+  compl := compl,
+  le_himp_iff := le_himp_iff,
+  himp_bot := λ a, sup_bot_eq,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a co-Heyting algebra from the difference alone. -/
+@[reducible] -- See note [reducible non instances]
+def coheyting_algebra.of_sdiff [distrib_lattice α] [bounded_order α] (sdiff : α → α → α)
+  (sdiff_le_iff : ∀ a b c, sdiff a b ≤ c ↔ a ≤ b ⊔ c) : coheyting_algebra α :=
+{ sdiff := sdiff,
+  hnot := λ a, sdiff ⊤ a,
+  sdiff_le_iff := sdiff_le_iff,
+  top_sdiff := λ a, rfl,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a co-Heyting algebra from the Heyting negation alone. -/
+@[reducible] -- See note [reducible non instances]
+def coheyting_algebra.of_hnot [distrib_lattice α] [bounded_order α] (hnot : α → α)
+  (sdiff_le_iff : ∀ a b c, a ⊓ hnot b ≤ c ↔ a ≤ b ⊔ c) : coheyting_algebra α :=
+{ sdiff := λ a b, (a ⊓ hnot b),
+  hnot := hnot,
+  sdiff_le_iff := sdiff_le_iff,
+  top_sdiff := λ a, top_inf_eq,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+section generalized_heyting_algebra
+variables [generalized_heyting_algebra α] {a b c d : α}
+
+-- `p → q → r ↔ p ∧ q → r`
```

Comment:
I like that you give an interpretation of these results. Could you add a few words describing how to understand these interpretations for people who are unfamiliar? Something like
```suggestion
-- In this section, we'll give interpretations of these results in the Heyting algebra model of intuitionistic logic,
-- where `≤` can be interpreted as "validates", `⇨` as "implies", `⊓` as "and", `⊔` as "or", `⊥` as "false" and `⊤` as "true".
-- See also `Prop.heyting_algebra`.

-- `p → q → r ↔ p ∧ q → r`
```

Reply:
Do you want me to distinguish "validates" and "implies" in these interpretations?

Reply:
In general these are not the same thing in all logics. Of course we have a deduction theorem for intuitionistic logic stating they are equivalent, so I guess it's fine to confuse them here (with perhaps a footnote saying so for the pedantic).